### ISSUE TITLE
MRefBuilder: $(TargetFramework) gets ignored while generating GenerateRefInfo.proj.

### ISSUE THREAD

**Issue by FlorianKugler42** [EWSoftware/SHFB#759]
I have a .csproj using multi-targeting with `<TargetFrameworks>net452;net472</TargetFrameworks>`. This project references different versions of a dependency based on `$(TargetFramework)`. So far so good.

Now I add a .shfbproj for documenting this project. When building this documentation, it fails with the following error:

```
MRefBuilder : error : Unresolved assembly reference: DependencyNET452 (DependencyNET452, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null) required by MultiTargetLibrary [C:\Projects\3rd\GitHub\EWSoftware\SHFB_issue_multitargeting\MultiTargetLibraryDocumentation\Help\Working\GenerateRefInfo.proj]
```

As far as I can see, this happens because the assembly reference in GenerateRefInfo.proj is generated with `$(TargetFramework)` being empty:

```xml
<HintPath>C:\Projects\3rd\GitHub\EWSoftware\SHFB_issue_multitargeting\MultiTargetLibrary\..\Dependency\bin\Debug\Dependency.dll</HintPath>
```

This should rather be:
```xml
<HintPath>C:\Projects\3rd\GitHub\EWSoftware\SHFB_issue_multitargeting\MultiTargetLibrarynet452\..\Dependency\bin\Debug\Dependencynet452.dll</HintPath>
```

[SHFB_issue_multitargeting.zip](https://github.com/EWSoftware/SHFB/files/3841444/SHFB_issue_multitargeting.zip)
[LastBuild.log](https://github.com/EWSoftware/SHFB/files/3841448/LastBuild.log)
[GenerateRefInfo.proj.txt](https://github.com/EWSoftware/SHFB/files/3841451/GenerateRefInfo.proj.txt)

Thanks for looking into this!

**Comment by EWSoftware** [comment]
I've fixed this so that it sets the target framework in the documentation source project when it sees that multi-targeting is being.  I'll be publishing an update this weekend that will contain this fix.

**Comment by FlorianKugler42** [comment]
First of all, thank you for reacting so quickly to this - you have a much better turn-around-time than me. Now, while the released fix does indeed solve the problem as presented, I have discovered a variation of the problem during my on-going project file refactoring. I don't know if I should open a new issue, re-open this one or what else, so for now I just add this as a comment here:

[SHFB_issue_multitargeting.zip](https://github.com/EWSoftware/SHFB/files/3939733/SHFB_issue_multitargeting.zip)
[LastBuild.log](https://github.com/EWSoftware/SHFB/files/3939749/LastBuild.log)

I have moved some property definitions out of the *.csproj and into Directory.Build.props, and SHFB behavior in this case seems to be inconsistent. While it does pick up Directory.Build.props, $(TargetFramework) (and maybe other properties) appear to be undefined at this point. So defining properties based on $(TargetFramework) inside Directory.Build.props does again lead to "MRefBuilder : error : Unresolved assembly reference".

I have included in the ZIP two possible work-arounds which I think illustrate the problem. Thanks again for your efforts, they are much appreciated!

**Comment by EWSoftware** [comment]
This was an odd one.  I made the following changes to the projects and code to get it to work:

- Removed the reference in the SHFB project.  It was already being used as the documentation source.  I don't think this affected anything but it wasn't needed.
- Set the build dependencies in the solution so that the MultiTargetLibrary.Doc project was dependent on the other three.  This ensures that they all build first and their output exists for SHFB to find.
- Removed the "Dependency." namespace/type prefix from the code in theClass1.cs file in the MultiTargetDependency project.

The last one was the key to getting it all to work but I don't know exactly why.  That namespace/type prefix does not exist and when viewing the code, it is flagged as an error.  However, the project did build and produce an assembly but shouldn't have given the error.  The Dependencies node for that project in Visual Studio showed a yellow warning indicator which meant it couldn't resolve a reference.  Once I removed the invalid prefix from the code, and rebuilt it, the warning indicator went away.  That change also allowed the help file project to build.

I've attached the modified files for comparison.

[IssuesFixes.zip](https://github.com/EWSoftware/SHFB/files/3952370/IssuesFixes.zip)

**Comment by FlorianKugler42** [comment]
Thanks for debugging!

> Removed the reference in the SHFB project. It was already being used as the documentation source. I don't think this affected anything but it wasn't needed.

I always include the `<ProjectReference>` so the dependencies can get resolved automatically. Is the `<DocumentationSource>` also establishing the dependency in regards to MsBuild? Or thinking the other way around, I'd personally find it very intuitive if by adding the `<ProjectReference>` alone the referenced project would work as a documentation source. Maybe with an additional meta-data attribute to turn that off. This would map perfectly to my mental model of what (Project)References in project files are used for. But this is probably more of a feature request / suggestion tangential to the main issue here.

> Set the build dependencies in the solution so that the MultiTargetLibrary.Doc project was dependent on the other three. This ensures that they all build first and their output exists for SHFB to find.

I have observed problems with build-dependency resolution in the context of multi-targeting and SDK-style projects independently from SHFB myself. So this might be actually part of the problem. But I tested this, and for the .shfbproj, the explicit dependencies do not help. I think mainly because the solution file already defines the 2 dependencies for MultiTargetLibrary.csproj.

> Removed the "Dependency." namespace/type prefix from the code in theClass1.cs file in the MultiTargetDependency project.

Yeah, I shouldn't have called _all_ of the classes "Class1" ;-) Sorry for the confusion. The "Dependency.Class1" is a different class than "MultiTargetLibrary.Class1", to which this of course resolves after removing the "Dependency." prefix. So that explains why MRefBuilder in this case doesn't need the DependencyNET452.dll anymore.

> That namespace/type prefix does not exist and when viewing the code, it is flagged as an error.
> The Dependencies node for that project in Visual Studio showed a yellow warning indicator which meant it couldn't resolve a reference.

I think this is because MultiTargetLibray does not use `<ProjectReference>` to reference the dependency projects, but `<Reference>` directly. So when you open the fresh solution, the referenced DLLs are not there yet. After the first build (or maybe after some VS-internal background build for IntelliSense?), the yellow icons and the red squiggles should disappear. This is to model my actual use case which uses 3rd party DLLs provided as is (no *.csproj for the dependencies).
